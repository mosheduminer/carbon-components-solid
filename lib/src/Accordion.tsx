import { Component, For, mergeProps, JSX } from "solid-js";
import { Dynamic } from "solid-js/web";
import { usePrefix } from "./internal/usePrefix";

export type AccordionProps = {
  align?: "start" | "end";
  class?: string;
  disabled?: boolean;
  size?: "sm" | "md" | "lg" | "xl";
} & JSX.HTMLAttributes<HTMLLIElement>;

export const Accordion: Component<AccordionProps> = (props) => {
  props = mergeProps({ align: "end" }, props);
  const prefix = usePrefix();

  return (
    <ul
      class={`${prefix}--accordion`}
      classList={{
        [props.class as string]: !!props.class,
        [`${prefix}--accordion--${props.align}`]: props.align !== undefined,
        [`${prefix}--accordion--${props.size}`]: props.size !== undefined,
      }}
    >
      {props.disabled ? (
        <For each={props.children as Component[]}>
          {(child) => <Dynamic component={child} />}
        </For>
      ) : (
        props.children
      )}
    </ul>
  );
};
