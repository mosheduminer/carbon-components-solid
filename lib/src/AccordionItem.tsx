import {
  Component,
  createSignal,
  splitProps,
  JSX,
  mergeProps,
  createEffect,
} from "solid-js";
import { settings } from "carbon-components";
import { ChevronRight } from "./icons/16";
import { match } from "./internal/keyboard/match";
import keys from "./internal/keyboard/keys";

const { prefix } = settings;

export type AccordionItemProps = {
  handleAnimationEnd?: JSX.EventHandler<HTMLLIElement, AnimationEvent>;
  class?: string;
  disabled?: boolean;
  onHeadingClick?: (props: { event: MouseEvent; isOpen: boolean }) => any;
  open?: boolean;
  title?: JSX.Element;
} & JSX.HTMLAttributes<HTMLLIElement>;

export const AccordionItem: Component<AccordionItemProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLLIElement>;
  [props, rest] = splitProps(props, [
    "class",
    "children",
    "disabled",
    "onHeadingClick",
    "open",
    "title",
    "handleAnimationEnd",
  ]);
  props = mergeProps({ open: false }, props);
  const [isOpen, setIsOpen] = createSignal(props.open);
  const [prevIsOpen, setPrevIsOpen] = createSignal(props.open);
  const [animation, setAnimation] = createSignal("");

  createEffect(() => {
    if (props.open !== prevIsOpen()) {
      setAnimation(isOpen() ? "collapsing" : "expanding");
      setIsOpen(props.open);
      setPrevIsOpen(props.open);
    }
  });

  function onClick(event: MouseEvent) {
    const nextValue = !isOpen();
    setAnimation(isOpen() ? "collapsing" : "expanding");
    setIsOpen(nextValue);
    if (props.onHeadingClick) {
      props.onHeadingClick({ isOpen: nextValue, event });
    }
  }

  // If the AccordionItem is open, and the user hits the ESC key, then close it
  function onKeyDown(event: KeyboardEvent) {
    if (isOpen() && match(event, keys.Escape)) {
      setIsOpen(false);
    }
  }

  function handleAnimationEnd(
    event: AnimationEvent & {
      currentTarget: HTMLLIElement;
      target: Element;
    }
  ) {
    if (props.handleAnimationEnd) {
      props.handleAnimationEnd(event);
    }
    setAnimation("");
  }

  return (
    <li
      class={`${prefix}--accordion__item`}
      classList={{
        [`${prefix}--accordion__item--active`]: isOpen(),
        [`${prefix}--accordion__item--${animation()}`]: !!animation(),
        [`${prefix}--accordion__item--disabled`]: props.disabled,
        [props.class!]: !!props.class,
      }}
      {...rest}
      onAnimationEnd={handleAnimationEnd}
    >
      <button
        disabled={props.disabled}
        aria-expanded={isOpen()}
        class={`${prefix}--accordion__heading`}
        onClick={(e) => onClick(e)}
        onKeyDown={onKeyDown}
        type="button"
      >
        <ChevronRight class={`${prefix}--accordion__arrow`} />
        <div class={`${prefix}--accordion__title`}>{props.title}</div>
      </button>
      <div class={`${prefix}--accordion__content`}>{props.children}</div>
    </li>
  );
};
