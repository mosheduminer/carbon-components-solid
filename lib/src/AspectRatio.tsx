import { Component, mergeProps, JSX, splitProps } from "solid-js";
import settings from "carbon-components/es/globals/js/settings";
import { Dynamic } from "solid-js/web";

const { prefix } = settings;

export type AspectRatioProps = {
  as?: Component | string;
  class?: string;
  ratio?: "16x9" | "9x16" | "2x1" | "1x2" | "4x3" | "3x4" | "1x1";
} & JSX.HTMLAttributes<HTMLElement>;

export const AspectRatio: Component<AspectRatioProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLElement>;
  [props, rest] = splitProps(props, ["class", "children", "as", "ratio"]);
  props = mergeProps({ as: "div", ratio: "1x1" }, props);
  return (
    <Dynamic
      component={props.as!}
      classList={{ [props.class!]: !!props.class }}
      class={`${prefix}--aspect-ratio ${prefix}--aspect-ratio--${props.ratio}`}
      {...rest}
    >
      {props.children}
    </Dynamic>
  );
};
