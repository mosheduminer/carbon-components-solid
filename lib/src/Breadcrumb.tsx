import { Component, JSX, splitProps } from "solid-js";
import { settings } from "carbon-components";

const { prefix } = settings;

export type BreadcrumbProps = {
  'aria-label'?: string;
  class?: string;
  noTrailingSlash?: boolean;
  ref?: HTMLElement;
} & JSX.HTMLAttributes<HTMLElement>;

export const Breadcrumb: Component<BreadcrumbProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLElement>;
  [props, rest] = splitProps(props, ["class", "children", "aria-label", "noTrailingSlash", "ref"]);

  return (
    <nav
      class={`${prefix}--breadcrumb`}
      classList={{ [`${prefix}--breadcrumb--no-trailing-slash`]: props.noTrailingSlash }}
      {...rest}
      ref={props.ref}>
      <ol classList={{[props.class!]: !!props.class}}>{props.children}</ol>
    </nav>
  )
}
