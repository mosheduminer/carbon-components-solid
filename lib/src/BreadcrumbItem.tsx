import {
  Component,
  Switch,
  Match,
  children,
  JSX,
  createEffect,
  splitProps,
} from "solid-js";
import { Dynamic } from "solid-js/web";
import { settings } from "carbon-components";
import { OverflowMenu } from "./OverflowMenu";
import { Link } from "./Link";
import { OverflowMenuHorizontal } from "./icons/32";

const { prefix } = settings;

export type BreadcrumbItemProps = {
  "aria-current"?: string | boolean;
  class?: string;
  href?: string;
  isCurrentPage: boolean;
  ref?: HTMLLIElement;
} & JSX.HTMLAttributes<HTMLLIElement>;

export const BreadcrumbItem: Component<BreadcrumbItemProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLLIElement>;
  [props, rest] = splitProps(props, [
    "aria-current",
    "class",
    "href",
    "isCurrentPage",
    "ref",
  ]);

  const classList = () => ({
    [`${prefix}--breadcrumb-item`]: true,
    [`${prefix}--breadcrumb-item--current`]:
      props.isCurrentPage && props["aria-current"] !== "page",
    [props.class!]: !!props.class,
  });
  const chlds = children(() => props.children) as () => JSX.Element;

  return (
    <Switch
      fallback={() => {
        createEffect(() => {
          (chlds()! as HTMLElement).classList.add(`${prefix}--link`);
          if (props["aria-current"]) {
            (chlds()! as HTMLElement).setAttribute("aria-current", "true");
          } else {
            (chlds()! as HTMLElement).removeAttribute("aria-current");
          }
        });
        return (
          <li classList={classList()} ref={props.ref} {...rest}>
            {chlds()}
          </li>
        );
      }}
    >
      <Match when={typeof chlds() === "string" && props.href}>
        <li classList={classList()} ref={props.ref} {...rest}>
          <Link href={props.href!} aria-current={props["aria-current"]}>
            {chlds()}
          </Link>
        </li>
      </Match>
      <Match when={typeof chlds() === "function"}>
        <li classList={classList()} {...rest}>
          <Dynamic
            component={chlds() as typeof OverflowMenu}
            menuOptionsClass={`${prefix}--breadcrumb-menu-options`}
            menuOffset={{ top: 10, left: 59 }}
            renderIcon={(props: JSX.HTMLAttributes<SVGSVGElement>) => (
              <OverflowMenuHorizontal
                class={`${prefix}--overflow-menu__icon`}
                {...props}
              />
            )}
          />
        </li>
      </Match>
    </Switch>
  );
};
