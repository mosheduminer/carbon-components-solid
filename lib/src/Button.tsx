import {
  Component,
  JSX,
  splitProps,
  createSignal,
  onCleanup,
  Show,
  mergeProps,
  onMount,
} from "solid-js";
import { matches } from "./internal/keyboard/match";
import keys from "./internal/keyboard/keys";
import { ButtonKindsType } from "./types";
import { composeEventHandlers } from "./internal/events";
import { Dynamic } from "solid-js/web";
import { createId } from "./internal/id";
import { usePrefix } from "./internal/usePrefix";

export type ButtonProps = {
  dangerDescription?: string;
  disabled?: boolean;
  role?: string;
  size?: "default" | "field" | "small" | "sm" | "md" | "lg" | "xl" | "2xl";
  renderIcon?: Component;
  hasIconOnly?: boolean;
  href?: string;
  iconDescription?: string;
  isExpressive?: boolean;
  isSelected?: boolean;
  tabIndex?: number;
  tooltipAlignment?: "start" | "center" | "end";
  tooltipPosition?: "top" | "start" | "bottom" | "left";
  type?: "button" | "reset" | "submit";
  ref?:
    | HTMLElement
    | HTMLButtonElement
    | ((el: HTMLElement | HTMLButtonElement) => void);
  kind?: ButtonKindsType;
} & JSX.HTMLAttributes<HTMLElement>;

const dangerButtonVariants = ["danger", "danger--tertiary", "danger--ghost"];

const buttonImage = (
  renderIcon?: Component<{ class?: string }>,
  iconDescription?: string
) => {
  const prefix = usePrefix();
  return (
    <Show when={renderIcon}>
      <Dynamic
        component={renderIcon}
        aria-label={iconDescription}
        class={`${prefix}--btn__icon`}
        aria-hidden="true"
      />
    </Show>
  );
};

export const Button: Component<ButtonProps> = (props) => {
  const prefix = usePrefix();
  let rest: Omit<JSX.HTMLAttributes<HTMLElement>, "ref">;
  [props, rest] = splitProps(props, [
    "children",
    "dangerDescription",
    "disabled",
    "role",
    "size",
    "renderIcon",
    "hasIconOnly",
    "href",
    "iconDescription",
    "isExpressive",
    "isSelected",
    "tabIndex",
    "tooltipAlignment",
    "tooltipPosition",
    "type",
    "onClick",
    "onBlur",
    "onFocus",
    "onMouseEnter",
    "onMouseLeave",
    "ref",
    "kind",
  ]);
  props = mergeProps(
    {
      tabIndex: 0,
      type: "button",
      disabled: false,
      kind: "primary",
      size: "default",
      dangerDescription: "danger",
      tooltipAlignment: "center",
      tooltipPosition: "top",
      isExpressive: false,
    },
    props
  );
  let tooltipRef: HTMLDivElement;
  let buttonRef: HTMLElement;
  let tooltipTimeout: number;
  const [allowTooltipVisibility, setAllowTooltipVisibility] =
    createSignal(false);
  const [isHovered, setIsHovered] = createSignal(false);
  const [isFocused, setIsFocused] = createSignal(false);
  const [classList, setClassList] = createSignal<{
    [key: string]: boolean | undefined;
  }>({
    [`${prefix}--btn`]: true,
    [`${prefix}--btn--sm`]:
      (props.size === "small" && !props.isExpressive) ||
      (props.size === "sm" && !props.isExpressive),
    [`${prefix}--btn--md`]:
      (props.size === "field" && !props.isExpressive) ||
      (props.size === "md" && !props.isExpressive),
    [`${prefix}--btn--lg`]: props.size === "xl",
    [`${prefix}--btn--xl`]: props.size === "2xl",
    [`${prefix}--btn--${props.kind}`]: !!props.kind,
    [`${prefix}--btn--disabled`]: props.disabled,
    [`${prefix}--btn--expressive`]: props.isExpressive,
    [`${prefix}--tooltip--hidden`]:
      props.hasIconOnly && !allowTooltipVisibility(),
    [`${prefix}--tooltip--visible`]: isHovered(),
    [`${prefix}--btn--icon-only`]: props.hasIconOnly,
    [`${prefix}--btn--selected`]:
      props.hasIconOnly && props.isSelected && props.kind === "ghost",
    [`${prefix}--tooltip__trigger`]: props.hasIconOnly,
    [`${prefix}--tooltip--a11y`]: props.hasIconOnly,
    [`${prefix}--btn--icon-only--${props.tooltipPosition}`]:
      props.hasIconOnly && !!props.tooltipPosition,
    [`${prefix}--tooltip--align-${props.tooltipAlignment}`]:
      props.hasIconOnly && !!props.tooltipAlignment,
  });
  const potentiallyCloseTooltips = (evt: Event) =>
    setClassList((list) => ({
      ...list,
      [`${prefix}--tooltip--hidden`]: evt.currentTarget !== buttonRef,
    }));
  const handleFocus = (evt: FocusEvent) => {
    if (props.hasIconOnly) {
      potentiallyCloseTooltips(evt);
      setIsHovered(!isHovered);
      setIsFocused(true);
      setAllowTooltipVisibility(true);
    }
  };
  const handleBlur = () => {
    if (props.hasIconOnly) {
      setIsHovered(false);
      setIsFocused(false);
      setAllowTooltipVisibility(false);
    }
  };
  const handleMouseEnter = (evt: MouseEvent) => {
    if (props.hasIconOnly) {
      setIsHovered(true);
      tooltipTimeout && clearTimeout(tooltipTimeout);

      if (evt.target === tooltipRef) {
        setAllowTooltipVisibility(true);
        return;
      }

      potentiallyCloseTooltips(evt);

      setAllowTooltipVisibility(true);
    }
  };
  const handleMouseLeave = () => {
    if (!isFocused() && props.hasIconOnly) {
      tooltipTimeout = setTimeout(() => {
        setAllowTooltipVisibility(false);
        setIsHovered(false);
      }, 100) as unknown as number;
    }
  };
  const potentiallyPreventClickEvent = (evt: MouseEvent) => {
    // Prevent clicks on the tooltip from triggering the button click event
    if (evt.target === tooltipRef) {
      evt.preventDefault();
      return;
    }
  };
  const handleEscKeyDown = (event: KeyboardEvent) => {
    if (matches(event, [keys.Escape])) {
      setAllowTooltipVisibility(false);
      setIsHovered(false);
    }
  };
  onMount(() => document.addEventListener("keydown", handleEscKeyDown));
  onCleanup(() => document.removeEventListener("keydown", handleEscKeyDown));
  const assistiveId = createId();

  const assistiveText = () => {
    if (props.hasIconOnly)
      return (
        <div
          ref={tooltipRef}
          onMouseEnter={handleMouseEnter}
          class={`${prefix}--assistive-text`}
        >
          {props.iconDescription}
        </div>
      );
    else if (dangerButtonVariants.includes(props.kind!))
      return (
        <span id={assistiveId} class={`${prefix}--visually-hidden`}>
          {props.dangerDescription}
        </span>
      );
    else return undefined;
  };

  return (
    <Show
      when={!props.href || props.disabled}
      fallback={
        <a
          tabIndex={props.tabIndex}
          classList={classList()}
          ref={props.ref as HTMLAnchorElement}
          onMouseEnter={composeEventHandlers([
            props.onMouseEnter as () => any,
            handleMouseEnter,
          ])}
          onMouseLeave={composeEventHandlers([
            props.onMouseLeave as () => any,
            handleMouseLeave,
          ])}
          onFocus={composeEventHandlers([
            props.onFocus as () => any,
            handleFocus,
          ])}
          onBlur={composeEventHandlers([props.onBlur as () => any, handleBlur])}
          onClick={composeEventHandlers([
            potentiallyPreventClickEvent,
            props.onClick as () => any,
          ])}
          {...rest}
        >
          {assistiveText()}
          {props.children}
          {buttonImage(props.renderIcon, props.iconDescription)}
        </a>
      }
    >
      <button
        tabIndex={props.tabIndex}
        classList={classList()}
        ref={props.ref as HTMLButtonElement}
        type={props.type}
        disabled={props.disabled}
        aria-describedby={
          dangerButtonVariants.includes(props.kind!) ? assistiveId : undefined
        }
        aria-pressed={
          props.hasIconOnly && props.kind === "ghost"
            ? props.isSelected
            : undefined
        }
        onMouseEnter={composeEventHandlers([
          props.onMouseEnter as () => any,
          handleMouseEnter,
        ])}
        onMouseLeave={composeEventHandlers([
          props.onMouseLeave as () => any,
          handleMouseLeave,
        ])}
        onFocus={composeEventHandlers([
          props.onFocus as () => any,
          handleFocus,
        ])}
        onBlur={composeEventHandlers([props.onBlur as () => any, handleBlur])}
        onClick={composeEventHandlers([
          potentiallyPreventClickEvent,
          props.onClick as () => any,
        ])}
        {...rest}
      >
        {assistiveText()}
        {props.children}
        {buttonImage(props.renderIcon, props.iconDescription)}
      </button>
    </Show>
  );
};
