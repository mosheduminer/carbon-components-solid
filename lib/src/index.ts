export * from "./Accordion";
export * from "./AccordionItem";
export * from "./AspectRatio";
export * from "./Breadcrumb";
export * from "./BreadcrumbItem";
export * from "./Button";
export * from "./ButtonSet";
export * from "./Checkbox";
export * from "./CodeSnippet";
export * from "./ComboBox";
export * from "./ComposedModal";
export * from "./ContentSwitch";
export * from "./ContentSwitcher";
export * from "./ContextMenu";
export * from "./Copy";
export * from "./CopyButton";
export * from "./DangerButton";
export * from "./DatePicker";
export * from "./DatePickerInput";
export * from "./FileUploader";
export * from "./Pagination";
export * from "./PaginationNav";
export * from "./ProgressBar";
export * from "./SkeletonText";
export * from "./Select";
export * from "./SelectItemGroup";
export * from "./SelectItem";
export * from "./StructuredList";
export * from "./Tag";
export * from "./TextInput";
