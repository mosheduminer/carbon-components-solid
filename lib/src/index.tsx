export * from "./Accordion";
export * from "./AccordionItem";
export * from "./AspectRatio";
export * from "./Breadcrumb";
export * from "./BreadcrumbItem";
export * from "./Button";
export * from "./ButtonSet";
export * from "./Checkbox";
export * from "./CodeSnippet";
export * from "./ComboBox";
export * from "./ComposedModal";
export * from "./ContentSwitch";
export * from "./ContentSwitcher";
export * from "./ContextMenu";
export * from "./Copy";
export * from "./CopyButton";
export * from "./DangerButton";
export * from "./DatePicker";
export * from "./DatePickerInput";
export * from "./FileUploader";
export * as icons from "./icons";
export * from "./SkeletonText";
export * from "./TextInput"
