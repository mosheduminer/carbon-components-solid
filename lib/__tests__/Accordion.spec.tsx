import { Accordion, AccordionItem } from '../src';
import { fireEvent } from '@testing-library/dom';
import userEvent from "@testing-library/user-event"
import { cleanup, render, screen } from 'solid-testing-library';

describe('Accordion', () => {
  afterEach(cleanup);

  it('should render', () => {
    const { container } = render(() =>
      <Accordion class="extra-class">
        <AccordionItem class="child" title="Heading A">
          Panel A
        </AccordionItem>
        <AccordionItem class="child" title="Heading B">
          Panel B
        </AccordionItem>
        <AccordionItem class="child" title="Heading C">
          Panel C
        </AccordionItem>
      </Accordion>
    );

    expect(container).toMatchSnapshot();
  });

  describe('automated verification testing', () => {
    it('should have no Axe violations', async () => {
      render(() =>
        <Accordion class="extra-class">
          <AccordionItem class="child" title="Heading A">
            Panel A
          </AccordionItem>
          <AccordionItem class="child" title="Heading B">
            Panel B
          </AccordionItem>
          <AccordionItem class="child" title="Heading C">
            Panel C
          </AccordionItem>
        </Accordion>
      );

      await expect(screen.getByText('Heading A'))
        //@ts-ignore
        .toHaveNoAxeViolations();

      // click to open
      fireEvent.click(screen.getByText('Heading A'));

      // test when open
      await expect(screen.getByText('Heading A'))
        //@ts-ignore
        .toHaveNoAxeViolations();
    });

    it('should have no Accessibility Checker violations', async () => {
      render(() =>
        <main>
          <Accordion class="extra-class">
            <AccordionItem class="child" title="Heading A">
              Panel A
            </AccordionItem>
            <AccordionItem class="child" title="Heading B">
              Panel B
            </AccordionItem>
            <AccordionItem class="child" title="Heading C">
              Panel C
            </AccordionItem>
          </Accordion>
        </main>
      );

      await expect(screen.getByText('Heading A'))
        //@ts-ignore
        .toHaveNoACViolations(
          'Accordion'
        );

      // click to open
      fireEvent.click(screen.getByText('Heading A'));

      // test when open
      await expect(screen.getByText('Heading A'))
        //@ts-ignore
        .toHaveNoACViolations(
          'Opened Accordion'
        );
    });
  });

  describe('basic keyboard accessibility testing', () => {
    it('should receive focus', async () => {
      render(() =>
        <Accordion class="extra-class">
          <AccordionItem class="child" title="Heading A">
            Panel A
          </AccordionItem>
          <AccordionItem class="child" title="Heading B">
            Panel B
          </AccordionItem>
          <AccordionItem class="child" title="Heading C">
            Panel C
          </AccordionItem>
        </Accordion>
      );

      userEvent.tab();
      await expect(document.getElementsByTagName('button')[0]).toHaveFocus();
    });

    it('should open with enter', async () => {
      render(() =>
        <Accordion class="extra-class">
          <AccordionItem class="child" title="Heading A">
            Panel A
          </AccordionItem>
          <AccordionItem class="child" title="Heading B">
            Panel B
          </AccordionItem>
          <AccordionItem class="child" title="Heading C">
            Panel C
          </AccordionItem>
        </Accordion>
      );

      // userEvent.type clicks the element passed before typing without skipClick option
      // see: https://github.com/testing-library/user-event#typeelement-text-options
      userEvent.type(screen.getByText('Heading A'), '{enter}', {
        skipClick: true,
      });

      expect(screen.getByText('Panel A')).toBeDefined();
    });

    it('should open with spacebar', async () => {
      render(() =>
        <Accordion class="extra-class">
          <AccordionItem class="child" title="Heading A">
            Panel A
          </AccordionItem>
          <AccordionItem class="child" title="Heading B">
            Panel B
          </AccordionItem>
          <AccordionItem class="child" title="Heading C">
            Panel C
          </AccordionItem>
        </Accordion>
      );

      // userEvent.type clicks the element passed before typing without skipClick option
      // see: https://github.com/testing-library/user-event#typeelement-text-options
      userEvent.type(screen.getByText('Heading A'), '{space}', {
        skipClick: true,
      });

      expect(screen.getByText('Panel A')).toBeDefined();
    });
  });
});