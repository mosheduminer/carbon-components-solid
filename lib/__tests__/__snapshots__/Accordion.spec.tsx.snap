// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render 1`] = `
<div>
  <ul
    class="bx--accordion extra-class bx--accordion--end"
  >
    <li
      class="bx--accordion__item child"
    >
      <button
        aria-expanded="false"
        class="bx--accordion__heading"
        type="button"
      >
        <svg
          class="bx--accordion__arrow"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
          />
        </svg>
        <div
          class="bx--accordion__title"
        >
          Heading A
        </div>
      </button>
      <div
        class="bx--accordion__content"
      >
        Panel A
      </div>
    </li>
    <li
      class="bx--accordion__item child"
    >
      <button
        aria-expanded="false"
        class="bx--accordion__heading"
        type="button"
      >
        <svg
          class="bx--accordion__arrow"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
          />
        </svg>
        <div
          class="bx--accordion__title"
        >
          Heading B
        </div>
      </button>
      <div
        class="bx--accordion__content"
      >
        Panel B
      </div>
    </li>
    <li
      class="bx--accordion__item child"
    >
      <button
        aria-expanded="false"
        class="bx--accordion__heading"
        type="button"
      >
        <svg
          class="bx--accordion__arrow"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 8L6 13 5.3 12.3 9.6 8 5.3 3.7 6 3z"
          />
        </svg>
        <div
          class="bx--accordion__title"
        >
          Heading C
        </div>
      </button>
      <div
        class="bx--accordion__content"
      >
        Panel C
      </div>
    </li>
  </ul>
</div>
`;
